text/0003-hash-indexed-object-store.md

```diff
@@ -0,0 +1,624 @@
+# Hash-indexed object store
+
+## Summary
+
+[summary]: #summary
+
+RFC PR: [casperlabs/rfcs#0003](https://github.com/casperlabs/rfcs/pull/0003)
+
+There is a variety of "objects" floating around in our codebase at the moment: Deploys, WASM blobs, Blocks, DAG nodes and others. By standardizing them into a [Git](https://git-scm.com)-inspired object model we can simplify our networking, gossiping, retrieval, and storage logic greatly.
+
+## Motivation
+
+[motivation]: #motivation
+
+One possible implementation of a mechanism for a new node joining the network is to provide it with a block and have it download all parents and dependencies up until the well-known genesis block.
+
+With the current node architecture, this approach would require quite a few component instances: Separate instantiations of the gossiper and fetcher components for each type involved, which at the minimum are block and deploy if not counting DAG nodes. Storage needs to namespace according to types and component developers need to remember which effect to call for each. During all of this, the core logic remains the same for all IO operations of these different types though.
+
+By unifying all these types, we can simplify all of our IO components drastically. Ensuring that all have a unique hash that includes a type-discriminating tag ensures that we can use the same namespace for identifiers.
+
+As a result, our IO layer will be greatly simplified, as well as its usage. Possible gains are easier snapshotting, DB handling and shipping data around in general. Speed-ups through external caching become very real (see [Future possibilities](#future-possibilities)).
+
+Taking this one step further, we can express the tree structure given through hash references independent of the objects type as well, moving more IO logic away from business components.
+
+## Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+(It is recommended to read the [prior art](#prior-art) section beforehand)
+
+Every *object* that our node handles has common properties, namely:
+
+* It can be serialized to binary storage.
+* It has a unique identifying hash that depends solely on its contents.
```

Comment:
> The implication here is that it depends upon all of its contents...

@Fraser999 

I believe this can be made to be true in all cases via aggregation, with the identifying hash of an aggregate being the hash of the hashes of its sub-elements.

Reply:
For something like a `Deploy` where it needs to be signed by all parties before it's given to the network, then I can see this being workable.  But for `Block`s which get stored and have further signatures appended as time passes, it'd be pretty convoluted I think.

The signatures would need to be of something like the current block hash (effectively the hash of everything in the block except the signatures themselves), but as we accumulate them into the block, the key under which it'd be stored will keep changing.

I imagine we'd be better keeping the block static under its block hash, and treating the signatures as separate metadata.  Ideally the changing metadata would also be stored under a fixed key (e.g. `hash(block_hash + "metadata")`), but that doesn't really fit well into the storage model proposed here I think.

Reply:
@Fraser999 Please have a look at the [Finality Signatures](https://github.com/marc-casperlabs/ceps/blob/hash-index-object-store/text/0003-hash-indexed-object-store.md#finality-signatures) part, in my opinion the scheme there should address all the issues mentioned here specifically for finality signatures.